### PR TITLE
feat: add video upload support for facebook

### PR DESCRIPTION
**Pull request by tmladek**
## Description

Adds support for Facebook video uploads

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tmladek** [comment]
Known issues: `attachments` is at the moment always presumed to be a list of Video IDs, triggering special handling (updating the video with `published=true` instead of creating an actual post). There's nothing about the ID that could be used to tell whether it's a video or potentially a photo.

- add a `type` param to `MediaAttachment`?
- change `attachments` to `videos` (as that's the only use right now anyway?)
- `GET` various endpoints to see which one returns without error, giving info about the type of the item?
- ?

File: grid/social-media/publish-post/maps/facebook.suma [hunk 1]
```diff
@@ -9,16 +9,22 @@ Publish post to Facebook profile (page)
...
     title = "profileId field is missing"
     detail = "Missing or invalid profileId, either page or group ID is expected"
   }
 
-  return map error if(media.length === 0 && !input.text && !input.link) {
-    title = "Missing text or link"
-    detail = "Either text or link fields must be non-empty for text posts"
+  return map error if (media.length === 0 && attachments.length === 0 && !input.text && !input.link) {
+    title = "Missing text, link or attachments"
+    detail = "Either text, link or attachments fields must be non-empty for text posts"
```
Comment (jnv) [inline grid/social-media/publish-post/maps/facebook.suma:22]: Nitpicking: this doesn't apply to just text posts, but _any_ posts

```suggestion
    detail = "Either text, link or attachments fields must be non-empty"
```

File: grid/social-media/publish-post/maps/facebook.suma [hunk 2]
```diff
@@ -9,16 +9,22 @@ Publish post to Facebook profile (page)
...
+  }
+
+  return map error if(attachments.length > 1) {
+    title = "Only one attachment is supported by Facebook."
```
Comment (jnv) [inline grid/social-media/publish-post/maps/facebook.suma:26]: nit

```suggestion
    title = "Too many attachments"
```

File: grid/social-media/upload-from-url/maps/facebook.test.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { publishVideoTest } from './upload-from-url';
+
+publishVideoTest('facebook', 'single video', {
+  caption: 'Single video from Superface Station.',
+  url: 'https://download.samplelib.com/mp4/sample-5s.mp4',
+  uploadType: 'video',
+});
+
+/*
+publishVideoTest('facebook', 'single reel', {
```
Comment (jnv) [inline grid/social-media/upload-from-url/maps/facebook.test.ts:10]: Commented out code

File: grid/social-media/upload-from-url/maps/facebook.suma [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+          case 'error':
+            return 'error';
+          case 'processing':
+            return 'inProgress'
+          default:
+            return null;
+        }
+      })();
+
+      map error if (state === null) {
```
Comment (jnv) [inline grid/social-media/upload-from-url/maps/facebook.suma:59]: I prefer having an object with mapped states over IIFE, but I understand that switch/case has more flexibility. For your consideraton:

```suggestion
      states = {
        ready: 'finished',
        error: 'error',
        processing: 'inProgress',
      }
      state = states[body.status.video_status]
      
      map error if (!state) {
```

**Review by jnv — COMMENTED** [review]
LGTM, few nitpicks.

Regarding the attachments issue, I like this option from your suggestion:

> change attachments to videos (as that's the only use right now anyway?)

However, I am not sure how it will affect usage with Twitter (for images) and LinkedIn (for documents).

Technically that's a breaking change, but I don't think anyone's using that including ourselves? Correct order would be to keep attachments field, document it as deprecated with hint to use videos instead, and use attachments if it's set and videos isn't.
If we figure out we still need attachments in the future, we can "undeprecate" it.

The documentation for videos field should mention the upload-from-url profile.

**Review by jnv — APPROVED** [review]

File: grid/social-media/upload-from-url/maps/facebook.suma (hunk 4 above)
Comment (tmladek) [reply grid/social-media/upload-from-url/maps/facebook.suma:59]: Good point, I think the `switch` statement is a leftover from when I had more logic in there, an object/map is cleaner.

Comment (tmladek) [reply grid/social-media/upload-from-url/maps/facebook.suma]: Well, except that the linter in `station` currently doesn't include the fix and so errors out :sweat_smile: The fix is already merged in `main` AFAIK, but upgrading the linter is beyond the scope of this PR I think

**Comment by tmladek** [comment]
>However, I am not sure how it will affect usage with Twitter (for images) and LinkedIn (for documents).

Right - well, I suppose we'd have to add (yet) another attribute for that (`documents`, `images`)... So I'd rather vote for adding a `type` field?

**Comment by jnv** [comment]
> Right - well, I suppose we'd have to add (yet) another attribute for that (documents, images)... So I'd rather vote for adding a type field?

That'd make it harder to use. I vote for `videos` with type of `[string]` and if we need to support more types, backward compat would be trivial.

**Comment by tmladek** [comment]
Okay, one last thing I noticed that I want to point out before I change it - we already use `attachment`s with `type`s in station, even in the same category of profiles: https://github.com/superfaceai/station/blob/main/grid/social-media/posts/profile.supr#L78

**Comment by jnv** [comment]
> Okay, one last thing I noticed that I want to point out before I change it - we already use `attachment`s with `type`s in station, even in the same category of profiles: https://github.com/superfaceai/station/blob/main/grid/social-media/posts/profile.supr#L78

That's good point, but I think it's a different use case (heh). That structure is for reading the data, here we optimize for writing / input. And you won't be feeding that structure from one profile to the other. In fact, having a similar structure under the same field name can be confusing.

**Pull request merged**